ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcInputFormat.java

```diff
@@ -2693,6 +2696,77 @@ public static TypeDescription getDesiredRowTypeDescr(Configuration conf,
     return result;
   }
 
+  /**
+   * Based on the file schema and the low level file includes provided in the SchemaEvolution instance, this method
+   * calculates which top level columns should be included i.e. if any of the nested columns inside complex types is
+   * required, then its relevant top level parent column will be considered as required (and thus the full subtree).
+   * Hive and LLAP currently only supports column pruning on the first level, thus we need to calculate this ourselves.
```

Comment:
So for ACID tables column pruning is not working, since the data is contained in the row struct?

Reply:
ACID is a special case, and I think it does. For ACID tables what Hive pushed down as projection is already a list of column indices within the row struct. So within this row struct for the first level columns, it should work.